include/sketch.h

```diff
@@ -18,35 +20,65 @@
  * raise an error.
  */
 class Sketch {
+public:
+  static double num_bucket_factor; // Factor for how many buckets there are in this sketch.
+  static vec_t n;                  // Length of the vector this is sketching.
+  static size_t num_elems;         // length of our actual arrays in number of elements
+  static size_t num_buckets;       // Portion of array length, number of buckets
+  static size_t num_guesses;       // Portion of array length, number of guesses
+
+  using SketchUniquePtr = std::unique_ptr<Sketch,std::function<void(Sketch*)>>;
+
+private:
   // Seed used for hashing operations in this sketch.
   const long seed;
-  // Length of the vector this is sketching.
-  const vec_t n;
-  // Factor for how many buckets there are in this sketch.
-  const double num_bucket_factor;
+  // pointers to buckets
+  vec_t*      bucket_a;
+  vec_hash_t* bucket_c;
 
-  // Buckets of this sketch.
-  // Length is bucket_gen(n, num_bucket_factor) * guess_gen(n).
-  // For buckets[i * guess_gen(n) + j], the bucket has a 1/2^j probability
-  // of containing an index.
-  std::vector<vec_t> bucket_a;
-  std::vector<vec_hash_t> bucket_c;
   // Flag to keep track if this sketch has already been queried.
   bool already_quered = false;
 
   FRIEND_TEST(SketchTestSuite, TestExceptions);
   FRIEND_TEST(EXPR_Parallelism, N10kU100k);
 
-public:
+  
+  // Buckets of this sketch.
+  // Length is bucket_gen(n, num_bucket_factor) * guess_gen(n).
+  // For buckets[i * guess_gen(n) + j], the bucket has a 1/2^j probability
+  // of containing an index. The first two are pointers into the buckets array.
+  char buckets[1];
+
   /**
    * Construct a sketch of a vector of size n
-   * @param n Length of the vector to sketch.
+   * @param n Length of the vector to sketch. (static variable)
    * @param seed Seed to use for hashing operations
-   * @param num_bucket_factor Factor to scale the number of buckets in this sketch
+   * @param num_bucket_factor Factor to scale the number of buckets in this sketch (static variable)
    */
-  Sketch(vec_t n, long seed, double num_bucket_factor = .5);
-  Sketch(const Sketch &old);
+  Sketch(long seed);
+  Sketch(long seed, std::fstream &binary_in);
+
+public:
+  static SketchUniquePtr makeSketch(const Sketch &old);
+  static SketchUniquePtr makeSketch(long seed);
+  static SketchUniquePtr makeSketch(long seed, std::fstream &binary_in);
+  static Sketch* makeSketch(void* loc, long seed);
+  static Sketch* makeSketch(void* loc, long seed, std::fstream &binary_in);
+  static Sketch* makeSketch(void* loc, long seed, double num_bucket_factor, std::fstream &binary_in);
 
+  inline static unsigned get_num_elems() { 
```

Comment:
num_elems is public, so either delete this function or make it private

Reply:
Okay